lib/Conversion/TritonGPUToLLVM/TritonGPUToLLVMBase.h

```diff
@@ -514,22 +514,49 @@ class ConvertTritonGPUOpToLLVMPatternBase {
     unsigned rank = shape.size();
     assert(rank == order.size());
     auto reordered = reorder(shape, order);
-    auto reorderedMultiDim = delinearize(rewriter, loc, linear, reordered);
+    SmallVector<Value> reorderedMultiDim(rank);
+    if (auto constantOp = linear.getDefiningOp<arith::ConstantOp>()) {
+      unsigned intVal =
+          constantOp.getValue().cast<IntegerAttr>().getValue().getSExtValue();
+      reorderedMultiDim = delinearize(rewriter, loc, intVal, reordered);
+    } else {
+      reorderedMultiDim = delinearize(rewriter, loc, linear, reordered);
+    }
     SmallVector<Value> multiDim(rank);
     for (unsigned i = 0; i < rank; ++i) {
       multiDim[order[i]] = reorderedMultiDim[i];
     }
     return multiDim;
   }
 
+  SmallVector<Value> delinearize(ConversionPatternRewriter &rewriter,
+                                 Location loc, unsigned linear,
+                                 ArrayRef<unsigned> shape) const {
+    unsigned rank = shape.size();
+    assert(rank > 0);
+    SmallVector<Value> multiDim(rank);
+    if (rank == 1) {
+      multiDim[0] = i32_val(linear % shape[0]);
+    } else {
+      unsigned remained = linear;
+      for (auto &&en : llvm::enumerate(shape.drop_back())) {
+        unsigned dimSize = en.value();
+        multiDim[en.index()] = i32_val(remained % dimSize);
+        remained = remained % dimSize;
+      }
+      multiDim[rank - 1] = i32_val(remained);
```

Comment:
Doesn't this need `% dimSize` as well ?

Reply:
I'll just rewrite the for loop. Thanks for pointing it out